drivers/SmartThings/zigbee-lock/src/init.lua

```diff
@@ -300,6 +303,88 @@ local function init(driver, device)
+  }
+
+  -- this is where we decide whether or not we need to delay our lock event because we've
+  -- observed it coming before the event (or we're starting to compute the timer)
+  local delay = device:get_field(DELAY_LOCK_EVENT) or 100
+  if (delay < MAX_DELAY) then
+    device.thread:call_with_delay(delay+.5, function ()
+      device:emit_event_for_endpoint(zb_rx.address_header.src_endpoint.value, LOCK_STATE[value.value])
+    end)
+  else
+    local socket = require "socket"
+    device:set_field(DELAY_LOCK_EVENT, socket.gettime())
+    device:emit_event_for_endpoint(zb_rx.address_header.src_endpoint.value, LOCK_STATE[value.value])
+  end
+end
+
+local lock_operation_event_handler = function(driver, device, zb_rx)
+  local event_code = zb_rx.body.zcl_body.operation_event_code.value
+  local source = zb_rx.body.zcl_body.operation_event_source.value
+  local OperationEventCode = require "st.zigbee.generated.zcl_clusters.DoorLock.types.OperationEventCode"
+  local METHOD = {
+    [0] = "keypad",
+    [1] = "command",
+    [2] = "manual",
+    [3] = "rfid",
+    [4] = "fingerprint",
+    [5] = "bluetooth"
+  }
+  local STATUS = {
+    [OperationEventCode.LOCK]           = capabilities.lock.lock.locked(),
+    [OperationEventCode.UNLOCK]         = capabilities.lock.lock.unlocked(),
+    [OperationEventCode.ONE_TOUCH_LOCK] = capabilities.lock.lock.locked(),
+    [OperationEventCode.KEY_LOCK]       = capabilities.lock.lock.locked(),
+    [OperationEventCode.KEY_UNLOCK]     = capabilities.lock.lock.unlocked(),
+    [OperationEventCode.AUTO_LOCK]      = capabilities.lock.lock.locked(),
+    [OperationEventCode.MANUAL_LOCK]    = capabilities.lock.lock.locked(),
+    [OperationEventCode.MANUAL_UNLOCK]  = capabilities.lock.lock.unlocked()
+  }
+  local event = STATUS[event_code]
+  if (event ~= nil) then
+    event["data"] = {}
+    if (event_code == OperationEventCode.AUTO_LOCK) then
+      event.data.method = "auto"
+    else
+      event.data.method = METHOD[source]
+    end
+    if (source == 0) then
+      local code_id = zb_rx.body.zcl_body.user_id.value
+      local code_name = "Code "..code_id
+      local lock_codes = device:get_field("lockCodes")
+      if (lock_codes ~= nil and
+          lock_codes[code_id] ~= nil) then
+        code_name = lock_codes[code_id]
+      end
+      event.data = {method = METHOD[0], codeId = code_id .. "", codeName = code_name}
+    end
+
+    -- if this is an event corresponding to a recently-received attribute report, we
+    -- want to set our delay timer for future lock attribute report events
+    if device:get_latest_state(
+        device:get_component_id_for_endpoint(zb_rx.address_header.src_endpoint.value),
+        capabilities.lock.ID,
+        capabilities.lock.lock.ID) == event.value.value then
+      local preceding_event_time = device:get_field(DELAY_LOCK_EVENT) or 0
+      local socket = require "socket"
+      local time_diff = socket.gettime() - preceding_event_time
```

Comment:
I see here that the actual time is used. Maybe just require `socket` once at the top rather than at separate places in the code?